<?xml version="1.0" ?>
<robot name="prius_1">
  <link name="chassis_floor">
    <inertial>
      <mass value="57.833"/>
      <origin rpy="0 0 0" xyz="0 0 0" />
      <inertia ixx="2.71553" ixy="0" ixz="0" iyy="68.9" iyz="0" izz="71.6063"/>
    </inertial>
    <visual name="v1">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="2.8808 0.75 0.030921"/>
      </geometry>
      <material name="white">
        <color rgba="1 1 1 1" />
      </material>
    </visual>
  </link>
  <link name="front_axle">
    <!-- <inertial>
      <mass value="0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial> -->
    <visual name="v1">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.061842 1.05132 0.030921"/>
      </geometry>
      <material name="green">
        <color rgba="0 1 0 1" />
      </material>
    </visual>
  </link>
  <link name="left_tie_rod_arm">
    <!-- <inertial>
      <mass value="0"/>
      <origin rpy="0  0  0" xyz="-0.17246 -0.08623  0"/>
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial> -->
    <visual name="__v2">
      <origin rpy="0 0 0.46364" xyz="-0.1086 -0.0543 0"/>
      <geometry>
        <box size="0.204637 0.061842 0.030921"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
    <visual name="__v1">
      <origin rpy="0  0  0" xyz="0       0.02228  0"/>
      <geometry>
        <box size="0.061842 0.140955 0.030921"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
  </link>
  <link name="left_hub">
    <!-- <inertial>
      <mass value="0"/>
      <origin rpy="0  0  0" xyz="0  0  0"/>
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial> -->
    <visual name="__visual">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.0618421" radius="0.0154605"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
  </link>
  <link name="tie_rod">
    <!-- <inertial>
      <mass value="0"/>
      <origin rpy="0  0  0" xyz="0     -0.4085  0"/>
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial> -->
    <visual name="v1">
      <origin rpy="0 0 0" xyz="0 -0.4085 0"/>
      <geometry>
        <box size="0.061842 0.878847 0.030921"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
  </link>
  <link name="left_wheel">
    <inertial>
      <mass value="14.28"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.499912" ixy="0" ixz="0" iyy="0.778439" iyz="0" izz="0.499912"/>
    </inertial>
    <collision name="c1">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.215" radius="0.323342"/>
      </geometry>
    </collision>
    <visual name="wheel">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.215" radius="0.323342"/>
      </geometry>
      <material name="gray">
        <color rgba="0.2 0.2 0.2 1" />
      </material>
    </visual>
    <visual name="line1">
      <origin rpy="0 1.570796326794896557998982 3.141592653589793115997963" xyz="0 0 -0.1075"/>
      <geometry>
        <box size="0.0025 0.0025 0.25"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1" />
      </material>
    </visual>
    <visual name="line2">
      <origin rpy="3.141592653589793115997963 1.570796326794896557998982 -1.570796326794896557998982" xyz="0 0 -0.1075"/>
      <geometry>
        <box size="0.0025 0.0025 0.25"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1" />
      </material>
    </visual>
  </link>
  <link name="right_tie_rod_arm">
    <!-- <inertial>
      <mass value="0"/>
      <origin rpy="0  0  0" xyz="-0.17246  0.08623  0"/>
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial> -->
    <visual name="v1">
      <origin rpy="0 0 0" xyz="0 -0.02228 0"/>
      <geometry>
        <box size="0.061842 0.140955 0.030921"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
    <visual name="v2">
      <origin rpy="0 0 -0.46364" xyz="-0.1086 0.0543 0"/>
      <geometry>
        <box size="0.204637 0.061842 0.030921"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
  </link>
  <link name="right_hub">
    <!-- <inertial>
      <mass value="0"/>
      <origin rpy="0  0  0" xyz="0  0  0"/>
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial> -->
    <visual name="v1">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.0618421" radius="0.0154605"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
  </link>
  <link name="right_wheel">
    <inertial>
      <mass value="14.28"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.499912" ixy="0" ixz="0" iyy="0.778439" iyz="0" izz="0.499912"/>
    </inertial>
    <collision  name="c1">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.215" radius="0.323342"/>
      </geometry>
    </collision>
    <visual name="v2">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.215" radius="0.323342"/>
      </geometry>
       <material name="gray">
        <color rgba="0.2 0.2 0.2 1" />
      </material>
    </visual>
    <visual name="v1">
      <origin rpy="0 1.570796326794896557998982 3.141592653589793115997963" xyz="0 0 0.1075"/>
      <geometry>
        <box size="0.0025 0.0025 0.25"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1" />
      </material>
    </visual>
    <visual name="v3">
      <origin rpy="3.141592653589793115997963 1.570796326794896557998982 -1.570796326794896557998982" xyz="0 0 0.1075"/>
      <geometry>
        <box size="0.0025 0.0025 0.25"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1" />
      </material>
    </visual>
  </link>
  <link name="rear_axle">
    <inertial>
      <mass value="0"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial>
    <visual name="v1">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.061842 1.38898 0.030921"/>
      </geometry>
    </visual>
  </link>
  <link name="left_wheel_rear">
    <inertial>
      <mass value="14.28"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.499912" ixy="0" ixz="0" iyy="0.778439" iyz="0" izz="0.499912"/>
    </inertial>
    <collision name="c1">
      <origin rpy="0  0  0" xyz="0  0  0"/>
      <geometry>
        <cylinder length="0.215" radius="0.323342"/>
      </geometry>
    </collision>
    <visual name="v3">
      <origin rpy="0 1.570796326794896557998982 3.141592653589793115997963" xyz="0 0 -0.1075"/>
      <geometry>
        <box size="0.0025 0.0025 0.25"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1" />
      </material>
    </visual>
    <visual name="v2">
      <origin rpy="3.141592653589793115997963 1.570796326794896557998982 -1.570796326794896557998982" xyz="0 0 -0.1075"/>
      <geometry>
        <box size="0.0025 0.0025 0.25"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1" />
      </material>
    </visual>
    <visual name="v1">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.215" radius="0.323342"/>
      </geometry>
      <material name="gray">
        <color rgba="0.2 0.2 0.2 1" />
      </material>
    </visual>
  </link>
  <link name="right_wheel_rear">
    <inertial>
      <mass value="14.28"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.499912" ixy="0" ixz="0" iyy="0.778439" iyz="0" izz="0.499912"/>
    </inertial>
    <collision name="c1">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.215" radius="0.323342"/>
      </geometry>
    </collision>
    <visual name="v3">
      <origin rpy="0 1.570796326794896557998982 3.141592653589793115997963" xyz="0 0 0.1075"/>
      <geometry>
        <box size="0.0025 0.0025 0.25"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1" />
      </material>
    </visual>
    <visual name="v1">
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.215" radius="0.323342"/>
      </geometry>
      <material name="gray">
        <color rgba="0.2 0.2 0.2 1" />
      </material>
    </visual>
    <visual name="v2">
      <origin rpy="3.141592653589793115997963 1.570796326794896557998982 -1.570796326794896557998982" xyz="0 0 0.1075"/>
      <geometry>
        <box size="0.0025 0.0025 0.25"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1" />
      </material>
    </visual>
  </link>
  <link name="body">
    <inertial>
      <mass value="5.42794"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.00216238" ixy="0" ixz="0" iyy="0.500375" iyz="0" izz="0.501672"/>
    </inertial>
    <visual name="v1">
      <origin rpy="0 0 0" xyz="-2.27 -0.91139 -0.21858"/>
      <geometry>
        <mesh filename="prius.dae" scale="1.0 1.0 1.0"/>
      </geometry>
    </visual>
  </link>

  <joint name="chassis_floor_axle" type="fixed">
    <parent link="chassis_floor"/>
    <child link="front_axle"/>
    <origin rpy="0 0 0" xyz="1.40948 0 -0.054984"/>
    <axis xyz="0 0 0"/>
    <limit effort="0" lower="0" upper="0" velocity="0"/>
  </joint>
  <joint name="steering" type="revolute">
    <parent link="front_axle"/>
    <child link="left_tie_rod_arm"/>
    <origin rpy="0 0 0" xyz="0 0.49473 -0.03092"/>
    <axis xyz="0 0 1"/>
    <limit effort="0.0" lower="-0.401426" upper="0.5" velocity="0.0"/>
  </joint>
  <joint name="left_hub" type="fixed">
    <parent link="left_tie_rod_arm"/>
    <child link="left_hub"/>
    <origin rpy="1.570796326794896557998982 0 0" xyz="0 0.12368 0.02406"/>
    <axis xyz="0  0  0"/>
    <limit effort="0" lower="0" upper="0" velocity="0"/>
  </joint>
  <joint name="left_pin" type="revolute">
    <parent link="left_tie_rod_arm"/>
    <child link="tie_rod"/>
    <origin rpy="0 0 0" xyz="-0.17246 -0.08623  0.03092"/>
    <axis xyz="0 0 1"/>
    <limit effort="0" lower="-1.5" upper="1.5" velocity="0"/>
  </joint>
  <joint name="left_wheel_joint" type="revolute">
    <parent link="left_hub"/>
    <child link="left_wheel"/>
    <origin rpy="0 0 0" xyz="0 0 -0.183582999999999940676787"/>
    <axis xyz="0 0 -1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <joint name="axle_tie_rod_arm" type="revolute">
    <parent link="front_axle"/>
    <child link="right_tie_rod_arm"/>
    <origin rpy="0 0 0" xyz="0 -0.49473 -0.03092"/>
    <axis xyz="0 0 1"/>
    <limit effort="0.0" lower="-0.401426" upper="0.5" velocity="0.0"/>
  </joint>
  <joint name="right_hub" type="fixed">
    <parent link="right_tie_rod_arm"/>
    <child link="right_hub"/>
    <origin rpy="1.570796326794896557998982 0 0" xyz="0 -0.12368 0.02406"/>
    <axis xyz="0  0  0"/>
    <limit effort="0" lower="0" upper="0" velocity="0"/>
  </joint>
  <joint name="right_wheel_joint" type="revolute">
    <parent link="right_hub"/>
    <child link="right_wheel"/>
    <origin rpy="0 0 0" xyz="0 0 0.18357"/>
    <axis xyz="0 0 -1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <joint name="chassis_rear_axle" type="fixed">
    <parent link="chassis_floor"/>
    <child link="rear_axle"/>
    <origin rpy="0 0 0" xyz="-1.40948 0 -0.054984"/>
    <axis xyz="0 0 0"/>
    <limit effort="0" lower="0" upper="0" velocity="0"/>
  </joint>
  <joint name="rear_left_wheel_joint" type="revolute">
    <parent link="rear_axle"/>
    <child link="left_wheel_rear"/>
    <origin rpy="1.570796326794896557998982  0 0" xyz="0 0.80199  0"/>
    <axis xyz="0  0 -1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <joint name="rear_right_wheel_joint" type="revolute">
    <parent link="rear_axle"/>
    <child link="right_wheel_rear"/>
    <origin rpy="1.570796326794896557998982 0 0" xyz="0 -0.80199  0"/>
    <axis xyz="0  0 -1"/>
    <limit effort="10000" velocity="10000"/>
  </joint>
  <joint name="body" type="fixed">
    <parent link="chassis_floor"/>
    <child link="body"/>
    <origin rpy="0  0  0" xyz="0 0 0.006857999999999975226927518"/>
    <axis xyz="0  0  0"/>
    <limit effort="0" lower="0" upper="0" velocity="0"/>
  </joint>

  <loop_joint name="right_pin" type="continuous">
    <axis xyz="0 0 1"/>
    <link1 link="right_tie_rod_arm" rpy="0 0 0" xyz="-0.17246 0.08622399999999996733635044  0.03092"/>
    <link2 link="tie_rod" rpy="0 0 0" xyz="0 -0.817006 0"/>
  </loop_joint>

  <transmission name="steering_drive" type="SimpleTransmission">
    <joint name="steering"/>
    <actuator name="steering"/>
  </transmission>
  <transmission name="front_left_drive" type="SimpleTransmission">
    <joint name="left_wheel_joint"/>
    <actuator name="left_wheel_joint"/>
  </transmission>
  <transmission name="front_right_drive" type="SimpleTransmission">
    <joint name="right_wheel_joint"/>
    <actuator name="right_wheel_joint"/>
  </transmission>
</robot>
